### PR TITLE
ci: support releases with branch-protection

### PR DESCRIPTION
**Pull request by v1v**
### What

Split the release workflow in a different set of actions:
1) Run a pre-release workflow
2) Then a GitHub Pull Request will be created with the versioning changes and slack notifiaction will be sent with the PR details.
3) When the PR is merged then 
4) Run the release workflow.

### Why

There is a hard requirement to review every single change through a Pull Request. Hence releases cannot commit directly to the repository.

### Implementation details

I split https://github.com/lerna/lerna/tree/main/libs/commands/publish#readme in two steps:
- `lerna version` to prepare the changes in a different branch and then create a PR with those changes
- `lerna publish` with some flags to avoid pushing to git https://github.com/lerna/lerna/tree/main/libs/commands/version#--no-git-tag-version and https://github.com/lerna/lerna/tree/main/libs/commands/version#--no-push

### Rationale

Rather than automating the `release` step when the PR is merged, I decided to keep it simple and have two workflows that need to be triggered manually. 

We can think about automating the execution of the second one when the PR is merged... but I didn't want to get to creative for the shake of simplicity.

### Test

https://github.com/v1v/test-2fa

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -78,21 +86,9 @@ jobs:
 
       - name: Publish the release
         env:
-          DRY_RUN: ${{ inputs.dry-run }}
+          DRY_RUN: "${{ inputs.dry-run }}"
         run: npm run ci:release
 
-      - name: Setup credentials
```
Comment (v1v) [inline .github/workflows/release.yml:84]: See ternary operation in the top-level env section

File: scripts/ci-release.mjs [hunk 2]
```diff
@@ -108,7 +108,7 @@ async function dryRunMode() {
   try {
     await execa(
       'npx',
-      ['lerna', 'publish', `--registry=${registryUrl}`, '--no-push', '--yes'],
+      ['lerna', 'publish', `--registry=${registryUrl}`, '--no-push', '--no-git-tag-version', '--no-changelog', '--yes'],
```
Comment (v1v) [inline scripts/ci-release.mjs:111]: lerna version is the creating the files/tags

File: scripts/ci-release.mjs [hunk 3]
```diff
@@ -140,12 +140,10 @@ async function prodMode() {
...
 
   try {
-    await execa('npx', ['lerna', 'publish', `--otp=${totpCode}`, '--yes'], {
-      stdin: process.stdin,
-      env: {
-        GH_TOKEN: githubToken
-      }
-    })
+    await execa('npx',
+      ['lerna', 'publish', `--otp=${totpCode}`, '--no-push', '--no-git-tag-version', '--no-changelog', '--yes'],
```
Comment (v1v) [inline scripts/ci-release.mjs:144]: `lerna version` is the creating the files/tags

**Comment by devcorpio** [comment]
I agree, I don't see any problem with having 2 manual steps (pre-release, and release). We can revisit this in the future if we believe that automating this more is essential.

So, If I understand, the flow will be something like this:

- prerelease step will create a PR where we will see the bumped version of every package + changelog, etc
- release step will create the tags and associate them to the latest commit of main (the merge commit of the PR) and will publish the packages, and will upload artifacts to the CDN.

Is this correct?

**Comment by v1v** [comment]
> Edit: I was checking the code, and I believe the tags will be created in the PR. The tags will need to be associated to the latest commit in main instead (which will be the merge commit of the PR).

Good point. When I tested this out I merged with rebase and things were ok:

- https://github.com/v1v/test-2fa/commit/21e6f1eebd3c3a159abc4f735be464034b63c677 points to https://github.com/v1v/test-2fa/releases/tag/test-2fa-foo%401.4.0-rc.13 and also to https://github.com/v1v/test-2fa/commit/9c7ac9a90ca3cedc225625d5042dd7ec386963b4

But I'm now a bit confused whether that behaviour might be different if using `squash and merge`. 

Let me explore whether I can change this to use `lerna version` without tagging and `lerna publish` with tagging.

File: scripts/ci-release.mjs (hunk 3 above)
Comment (v1v) [reply scripts/ci-release.mjs:144]: https://github.com/lerna/lerna/blob/v4.0.0/commands/version/README.md#--no-git-tag-version
https://github.com/lerna/lerna/blob/v4.0.0/commands/version/README.md#--no-changelog

**Comment by v1v** [comment]
> Let me explore whether I can change this to use lerna version without tagging and lerna publish with tagging.

https://github.com/elastic/apm-agent-rum-js/pull/1458/commits/cdc989ea38c5867155aae742289d011cf2cad384 is the one to avoid tagging during the `pre-release` workflow. The `post-release` workflow then will be the one pushing the changes and creating the release.

For example, in my playground, I ran:

* https://github.com/v1v/test-2fa/actions/runs/7181174251 and
  * https://github.com/v1v/test-2fa/pull/60 was created
* Then I ran https://github.com/v1v/test-2fa/actions/runs/7181227373
  * https://github.com/v1v/test-2fa/releases/tag/test-2fa-bar%402.0.1 was created

There is one thing that I could not figure out and was the one regarding tagging the package named https://github.com/v1v/test-2fa/tree/master/packages/test-2fa-foo for some reason it only tagged the main package called https://github.com/v1v/test-2fa/tree/master/packages/test-2fa-bar

I don't know whether tagging all the packages is a hard requirement.

@devcorpio , I need your thoughts

**Comment by vigneshshanmugam** [comment]
> for some reason it only tagged the main package called https://github.com/v1v/test-2fa/tree/master/packages/test-2fa-bar

Does both of the package had changes? Can you point us to the commit where this happened? Thanks.

**Comment by v1v** [comment]
> Does both of the packages had changes? Can you point us to the commit where this happened? Thanks.

No code changes in any of those packages. Aha, that might be the reason.

I've just tried some code changes with https://github.com/v1v/test-2fa/pull/61

Then I ran [the prepare-release](https://github.com/v1v/test-2fa/actions/runs/7192166789) produced https://github.com/v1v/test-2fa/pull/62
And after merging it I ran [the release](https://github.com/v1v/test-2fa/actions/runs/7192174792), so [test-2fa-bar@2.1.0](https://github.com/v1v/test-2fa/releases/tag/test-2fa-bar%402.1.0) was created but no tag/releases for the package `test-2fa-foo`.


IIUC, `lerna version` is the one producing all the relevant tags but when using `--no-push` no tags are created at all. And `lerna publish` does not know anything about those tags, that's managed by `lerna version`.

Finally, I'll need to explore more this workaround, somehow `lerna pubish` does publish a new version, which it's incorrect

image

while I'd have expected the same versions as defined in https://github.com/v1v/test-2fa/pull/62

I'll transition this PR to draft

**Comment by v1v** [comment]
> Finally, I'll need to explore more this workaround, somehow lerna pubish does publish a new version, which it's incorrect

I followed the steps explained in https://github.com/lerna/lerna/issues/1957#issuecomment-476894990 and it worked like a charm:

https://github.com/v1v/test-2fa/pull/65 was created and when merged then the `lerna publish` kept those package versions:

```
Found 2 packages to publish:
 - test-2fa-bar => 2.1.3
 - test-2fa-foo => 3.4.0
```

And publish them in npmjs:

image

File: scripts/ci-release.mjs (hunk 2 above)
Comment (v1v) [reply scripts/ci-release.mjs:111]:
```suggestion
      ['lerna', 'publish', 'from-package', `--registry=${registryUrl}`, '--no-push', '--no-git-tag-version', '--no-changelog', '--yes'],
```

File: scripts/ci-release.mjs (hunk 3 above)
Comment (v1v) [reply scripts/ci-release.mjs:144]:
```suggestion
      ['lerna', 'publish', 'from-package', `--otp=${totpCode}`, '--no-push', '--no-git-tag-version', '--no-changelog', '--yes'],
```

**Comment by devcorpio** [comment]
apologies for the delay, this afternoon/evening I will add the results of a few tests (mainly with the "tag strategy it follows") I'm doing with the current Lerna setup 

But from what I see of the new process, everything looks awesome.

**Comment by devcorpio** [comment]
> apologies for the delay, this afternoon/evening I will add the results of a few tests I'm doing with the current (the "old" one) Lerna setup

**1. Changing a package which no other package depends on (e.g. apm-rum-react)**

Result: Lerna only versions the mentioned package.

Screenshot 2023-12-13 at 15 49 35

**2. Changing a package which other packages depend on (e.g. apm-rum-core)**

Screenshot 2023-12-13 at 15 51 32

Result: Lerna versions the package I modified and the ones that depend on the modified package.

**Conclusion:**

The behaviour you [see](https://github.com/elastic/apm-agent-rum-js/pull/1458#issuecomment-1853486830) in the tests you are doing is correct. 🚀  One last test that you might do (if you have time) is to modify one of your packages to depend on another one, and then see if the behaviour that I just described (case 2) is reproduced, too

**Comment by vigneshshanmugam** [comment]
Great work @v1v 🎉 

One more thing I would like to check is whether our publishing for single package works after this change? https://github.com/elastic/apm-agent-rum-js/blob/7cd5f5ecbe02ee4ddc3e648927fa4ac0ce3dbb28/package.json#L14

Its not a huge deal if we cant make it work and not a requirement unless we need a real emergency situation. But just want to bring it to your notice.

**Comment by v1v** [comment]
> One more thing I would like to check is whether our publishing for single package works after this change?

Pushing directly to the main branch is not allowed; therefore the script will not work with the default lerna.json.

**Comment by v1v** [comment]
> One last test that you might do (if you have time) is to modify one of your packages to depend on another one and then see if the behaviour that I just described (case 2) is reproduced, too

I'll test this out and update this comment with the outcome.

**Review by devcorpio — APPROVED** [review]
LGTM! 🎉 

this week (perhaps the next one) we will do our first release with this system. We want to release the changes of [this](https://github.com/elastic/apm-agent-rum-js/pull/1462) PR (once it's approved & merged)

**Comment by v1v** [comment]
> this week (perhaps the next one) we will do our first release with this system. We want to release the changes of https://github.com/elastic/apm-agent-rum-js/pull/1462 PR (once it's approved & merged)

Awesome! ❤️ 

I'll be on Christmas break for two weeks starting early next week. Feel free to merge this PR and run the first release with it, if something goes terribly wrong @amannocci is around, so feel free to reach him. Or, if needed, please feel free to revert this PR or wait until I'm back, whatever it suits you.

Other than that, have a lovely 🎄 time 🙇

**Comment by devcorpio** [comment]
> Awesome! ❤️
> 
> I'll be on Christmas break for two weeks starting early next week. Feel free to merge this PR and run the first release with it, if something goes terribly wrong @amannocci is around, so feel free to reach him. Or, if needed, please feel free to revert this PR or wait until I'm back, whatever it suits you.
> 
> Other than that, have a lovely 🎄 time 🙇

thank you! Have a lovely 🎄  you too 🎉

**Comment by devcorpio** [comment]
excellent! The [release](https://github.com/elastic/apm-agent-rum-js/releases/tag/%40elastic%2Fapm-rum%405.16.0) process worked!  🎉 🥳 

- Pre-release job: https://github.com/elastic/apm-agent-rum-js/actions/runs/7337735981/job/19979110773
- Release PR: https://github.com/elastic/apm-agent-rum-js/pull/1465
- Release job: https://github.com/elastic/apm-agent-rum-js/actions/runs/7337869008/job/19979457000

--
At the same time, I found two things that we should take into account:

- The release process has only created one tag `@elastic/apm-rum@5.16.0`. I needed to create the other ones manually: `@elastic/apm-rum-vue@2.1.6`, `@elastic/apm-rum-react@2.0.2`, `@elastic/apm-rum-core@5.21.0`, `@elastic/apm-rum-angular@3.0.3`.

Example:

Screenshot tag example

- The pre-release action doesn't work well with [dry-run mode](https://github.com/elastic/apm-agent-rum-js/actions/runs/7337715651/job/19979058196) (not a big deal, since we can see how the versions will be once the PR is created, but I just wanted to share).

**Pull request merged**